open-apparel-registry/open-apparel-registry
# Add Facility delete API

## Overview

Add a superuser-only API for deleting facilities and updating any related matches.

Connects #597

## Demo

![2019-06-19 13 08 18](https://user-images.githubusercontent.com/17363/59797116-cf765f80-9293-11e9-8132-fd7cb6eb9d68.gif)


### Delete method does not appear in the API docs

<img width="1018" alt="Screen Shot 2019-06-19 at 1 11 23 PM" src="https://user-images.githubusercontent.com/17363/59797099-c1c0da00-9293-11e9-927b-ff7350342140.png">

### Deleted items are searchable on the list details page

<img width="1181" alt="Screen Shot 2019-06-19 at 1 14 48 PM" src="https://user-images.githubusercontent.com/17363/59797313-4875b700-9294-11e9-9cb0-258dabfaaa17.png">


## Notes

Redirection using `FacilityAlias` is not implemented in this PR.

To allow for deletes we need to convert from using `ReadOnlyModelViewSet` to manually opting in to the list, retrieve, and delete methods.

We will be reserving this action for superusers, so we exclude it from the public API documentation by defining a custom `AutoSchema` subclass and overriding the appropriate method.

Adds a `DELETED` status for `FacilityListItem`s so we can correctly display and filter for them on the contributor facility list detail page.

Adds a `DELETE_FACILITY` processing action type. The delete operation appends an entry to the `processing_results` array on the `FacilityListItem` that includes the ID of the deleted `Facility`. This is not strictly required, but continues the theme of recording all the events related to the `FacilityListItem`.

Adds helper methods to `Facility` which fetch `FacilityMatch` records to help decrease the size of the `destroy` view function.

Returns a bad request response if there is an approved `FacilityClaim` object related to the `Facility`. We want to force superusers to revoke an existing claim before deleting a `Facility` to trigger a notification to the claimant.

Change foreign keys on FacilityMatch from CASCADE to PROTECT. Whenever we delete a `FacilityMatch` row we want to use the `changeReason` attribute provided by `django-simple-history`. Setting `PROTECT` on the relationship requires us to manually manage our delete process.

## Testing Instructions

* Test that migrations apply to an older DB version
  * `./scripts/manage migrate`
* Reset the DB
  * `./scripts/manage resetdb`
  * `./scripts/manage loadfixtures`
  * `./scripts/manage processfixtures`
* Log in as `c1@example.com` and generate an API key on the profile page. Save this for later.
* Log in as `c15@example.com`
* Browse http://localhost:6543/lists/15, search for `Regina Miracle Intimate` and verify that the status is `MATCHED`.
* On the map page, search for `Regina Miracle Intimate`. Verify that the facility has three contributors.
* Copy the OAR ID and submit a delete API request 
```
curl --request DELETE \
  --url http://localhost:8081/api/facilities/{{OAR ID}} \
  --header 'authorization: Token {{API KEY}}'
```
* Refresh the facility details page and verify it says `Not Found`
* Search for `Regina Miracle Intimate` and verify that the facility now has only 2 contributors.
* Browse http://localhost:6543/lists/15 and verify that the item now has a status of `NEW_FACILITY`

## Checklist
